[pushover] Feature request: Send an Image URL directly

> - Sending a PNG image from internet: `sendPushoverImage("Test", "https://jaxenter.de/wp-content/uploads/2017/12/openhablogo.png")`
> 
> Doesn't work. But would be a very nice feature. Can you please add it. Thanks.
> ```
> 2018-03-29 12:18:30.558 [ERROR] [ab.action.pushover.internal.Pushover] - An error occurred while notifying your mobile device.
> java.io.FileNotFoundException: File is not a normal file.
> 	at org.apache.commons.httpclient.methods.multipart.FilePartSource.<init>(FilePartSource.java:69) ~[?:?]
> 	at org.apache.commons.httpclient.methods.multipart.FilePartSource.<init>(FilePartSource.java:89) ~[?:?]
> 	at org.apache.commons.httpclient.methods.multipart.FilePart.<init>(FilePart.java:160) ~[?:?]
> 	at org.openhab.action.pushover.internal.Pushover.pushover(Pushover.java:368) ~[?:?]
> 	at org.openhab.action.pushover.internal.Pushover.sendPushoverImage(Pushover.java:107) ~[?:?]
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
> 	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.invokeOperation(XbaseInterpreter.java:1085) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.invokeOperation(XbaseInterpreter.java:1060) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._invokeFeature(XbaseInterpreter.java:1046) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.invokeFeature(XbaseInterpreter.java:991) ~[?:?]
> 	at org.eclipse.smarthome.model.script.interpreter.ScriptInterpreter.invokeFeature(ScriptInterpreter.java:143) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._doEvaluate(XbaseInterpreter.java:901) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._doEvaluate(XbaseInterpreter.java:864) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.doEvaluate(XbaseInterpreter.java:223) ~[?:?]
> 	at org.eclipse.smarthome.model.script.interpreter.ScriptInterpreter.doEvaluate(ScriptInterpreter.java:219) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.internalEvaluate(XbaseInterpreter.java:203) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._doEvaluate(XbaseInterpreter.java:446) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.doEvaluate(XbaseInterpreter.java:227) ~[?:?]
> 	at org.eclipse.smarthome.model.script.interpreter.ScriptInterpreter.doEvaluate(ScriptInterpreter.java:219) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.internalEvaluate(XbaseInterpreter.java:203) ~[?:?]
> 	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.evaluate(XbaseInterpreter.java:189) ~[?:?]
> 	at org.eclipse.smarthome.model.script.runtime.internal.engine.ScriptImpl.execute(ScriptImpl.java:82) ~[?:?]
> 	at org.eclipse.smarthome.model.rule.runtime.internal.engine.RuleEngineImpl.lambda$2(RuleEngineImpl.java:339) ~[?:?]
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
> 	at java.lang.Thread.run(Thread.java:748) [?:?]
> ```

Example for current API: `sendPushoverMessage(pushoverBuilder("Test").withAttachment("https://jaxenter.de/wp-content/uploads/2017/12/openhablogo.png"))`

See https://github.com/openhab/openhab1-addons/pull/5527#discussion_r178014252 and https://community.openhab.org/t/pushover-image-support/39201/53.

--- a/bundles/org.openhab.binding.pushover/README.md
+++ b/bundles/org.openhab.binding.pushover/README.md
@@ -46,7 +46,7 @@ One has to pass a `null` value if it should be skipped or the default value for 
 
 - `sendMonospaceMessage(String message, @Nullable String title)` - This method is used to send a monospace message.
 
-- `sendAttachmentMessage(String message, @Nullable String title, String attachment, @Nullable String contentType)` - This method is used to send a message with an attachment. It takes a local path or URL to the attachment (parameter `attachment` **mandatory**) and an optional `contentType` to define the content-type of the attachment (default: `image/jpeg`).
+- `sendAttachmentMessage(String message, @Nullable String title, String attachment, @Nullable String contentType)` - This method is used to send a message with an attachment. It takes a local path or URL to the attachment (parameter `attachment` **mandatory**). Additionally you can pass a data URI scheme to this parameter. Optionally pass a `contentType` to define the content-type of the attachment (default: `image/jpeg` or guessed from image data).
 
 - `sendURLMessage(String message, @Nullable String title, String url, @Nullable String urlTitle)` - This method is used to send a message with an URL. A supplementary `url` to show with the message and a `urlTitle` for the URL, otherwise just the URL is shown.
 
@@ -74,6 +74,16 @@ demo.rules:
 val actions = getActions("pushover", "pushover:pushover-account:account")
 // send HTML message
 actions.sendHtmlMessage("Hello <font color='green'>World</font>!", "openHAB")
+```
+
+```java
+val actions = getActions("pushover", "pushover:pushover-account:account")
+// send message with attachment
+actions.sendAttachmentMessage("Hello World!", "openHAB", "/path/to/my-local-image.png", "image/png")
+actions.sendAttachmentMessage("Hello World!", "openHAB", "https://www.openhab.org/openhab-logo-square.png", null)
+actions.sendAttachmentMessage("Hello World!", "openHAB", "data:[<media type>][;base64],<data>", null)
+// in case you want to send the content of an Image Item (RawType)
+actions.sendAttachmentMessage("Hello World!", "openHAB", myImageItem.state.toFullString, null)
 ```
 
 ```java

--- a/bundles/org.openhab.binding.pushover/README.md
+++ b/bundles/org.openhab.binding.pushover/README.md
@@ -46,7 +46,7 @@ One has to pass a `null` value if it should be skipped or the default value for 
 
 - `sendMonospaceMessage(String message, @Nullable String title)` - This method is used to send a monospace message.
 
-- `sendAttachmentMessage(String message, @Nullable String title, String attachment, @Nullable String contentType)` - This method is used to send a message with an attachment. It takes a (local) path to the attachment (parameter `attachment` **mandatory**) and an optional `contentType` to define the content-type of the attachment (default: `image/jpeg`).
+- `sendAttachmentMessage(String message, @Nullable String title, String attachment, @Nullable String contentType)` - This method is used to send a message with an attachment. It takes a local path or URL to the attachment (parameter `attachment` **mandatory**) and an optional `contentType` to define the content-type of the attachment (default: `image/jpeg`).
 
 - `sendURLMessage(String message, @Nullable String title, String url, @Nullable String urlTitle)` - This method is used to send a message with an URL. A supplementary `url` to show with the message and a `urlTitle` for the URL, otherwise just the URL is shown.
 

--- a/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverConfigurationException.java
+++ b/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverConfigurationException.java
@@ -20,7 +20,7 @@ import org.eclipse.jdt.annotation.NonNullByDefault;
  * @author Christoph Weitkamp - Initial contribution
  */
 @NonNullByDefault
-public class PushoverConfigurationException extends RuntimeException {
+public class PushoverConfigurationException extends IllegalArgumentException {
 
     private static final long serialVersionUID = 1L;
 

--- a/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java
+++ b/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java
@@ -14,7 +14,8 @@ package org.openhab.binding.pushover.internal.connection;
 
 import java.io.File;
 import java.io.IOException;
-import java.util.Arrays;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.List;
 import java.util.stream.Collectors;
 
@@ -24,6 +25,8 @@ import org.eclipse.jetty.client.api.ContentProvider;
 import org.eclipse.jetty.client.util.MultiPartContentProvider;
 import org.eclipse.jetty.client.util.PathContentProvider;
 import org.eclipse.jetty.client.util.StringContentProvider;
+import org.openhab.core.io.net.http.HttpUtil;
+import org.openhab.core.library.types.RawType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -55,7 +58,7 @@ public class PushoverMessageBuilder {
     private static final int MAX_MESSAGE_LENGTH = 1024;
     private static final int MAX_TITLE_LENGTH = 250;
     private static final int MAX_DEVICE_LENGTH = 25;
-    private static final List<Integer> VALID_PRIORITY_LIST = Arrays.asList(-2, -1, 0, 1, 2);
+    private static final List<Integer> VALID_PRIORITY_LIST = List.of(-2, -1, 0, 1, 2);
     private static final int DEFAULT_PRIORITY = 0;
     public static final int EMERGENCY_PRIORITY = 2;
     private static final int MIN_RETRY_SECONDS = 30;
@@ -239,18 +242,36 @@ public class PushoverMessageBuilder {
         }
 
         if (attachment != null) {
-            File file = new File(attachment);
-            if (!file.exists()) {
-                throw new IllegalArgumentException(
-                        String.format("Skip sending the message as file '%s' does not exist.", attachment));
-            }
-            try {
-                body.addFilePart(MESSAGE_KEY_ATTACHMENT, file.getName(),
-                        new PathContentProvider(contentType, file.toPath()), null);
-            } catch (IOException e) {
-                logger.debug("IOException occurred - skip sending message: {}", e.getLocalizedMessage(), e);
-                throw new PushoverCommunicationException(
-                        String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+            if (attachment.startsWith("http")) {
+                RawType content = HttpUtil.downloadImage(attachment, 10000);
+                if (content == null) {
+                    throw new IllegalArgumentException(
+                            String.format("Skip sending the message as content '%s' does not exist.", attachment));
+                }
+                try {
+                    Path tmpFile = Files.createTempFile("pushover-", ".tmp");
+                    Files.write(tmpFile, content.getBytes());
+                    body.addFilePart(MESSAGE_KEY_ATTACHMENT, tmpFile.toFile().getName(),
+                            new PathContentProvider(contentType, tmpFile), null);
+                } catch (IOException e) {
+                    logger.debug("IOException occurred - skip sending message: {}", e.getLocalizedMessage(), e);
+                    throw new PushoverCommunicationException(
+                            String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+                }
+            } else {
+                File file = new File(attachment);
+                if (!file.exists()) {
+                    throw new IllegalArgumentException(
+                            String.format("Skip sending the message as file '%s' does not exist.", attachment));
+                }
+                try {
+                    body.addFilePart(MESSAGE_KEY_ATTACHMENT, file.getName(),
+                            new PathContentProvider(contentType, file.toPath()), null);
+                } catch (IOException e) {
+                    logger.debug("IOException occurred - skip sending message: {}", e.getLocalizedMessage(), e);
+                    throw new PushoverCommunicationException(
+                            String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+                }
             }
         }
 

--- a/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java
+++ b/bundles/org.openhab.binding.pushover/src/main/java/org/openhab/binding/pushover/internal/connection/PushoverMessageBuilder.java
@@ -79,7 +79,7 @@ public class PushoverMessageBuilder {
     private @Nullable String urlTitle;
     private @Nullable String sound;
     private @Nullable String attachment;
-    private String contentType = DEFAULT_CONTENT_TYPE;
+    private @Nullable String contentType;
     private boolean html = false;
     private boolean monospace = false;
 
@@ -242,21 +242,23 @@ public class PushoverMessageBuilder {
         }
 
         if (attachment != null) {
-            if (attachment.startsWith("http")) {
-                RawType content = HttpUtil.downloadImage(attachment, 10000);
-                if (content == null) {
+            String localAttachment = attachment;
+            if (localAttachment.startsWith("http")) { // support data HTTP(S) scheme
+                RawType rawImage = HttpUtil.downloadImage(attachment, 10000);
+                if (rawImage == null) {
                     throw new IllegalArgumentException(
                             String.format("Skip sending the message as content '%s' does not exist.", attachment));
                 }
+                addFilePart(createTempFile(rawImage.getBytes()),
+                        contentType == null ? rawImage.getMimeType() : contentType);
+            } else if (localAttachment.startsWith("data:")) { // support data URI scheme
                 try {
-                    Path tmpFile = Files.createTempFile("pushover-", ".tmp");
-                    Files.write(tmpFile, content.getBytes());
-                    body.addFilePart(MESSAGE_KEY_ATTACHMENT, tmpFile.toFile().getName(),
-                            new PathContentProvider(contentType, tmpFile), null);
-                } catch (IOException e) {
-                    logger.debug("IOException occurred - skip sending message: {}", e.getLocalizedMessage(), e);
-                    throw new PushoverCommunicationException(
-                            String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+                    RawType rawImage = RawType.valueOf(localAttachment);
+                    addFilePart(createTempFile(rawImage.getBytes()),
+                            contentType == null ? rawImage.getMimeType() : contentType);
+                } catch (IllegalArgumentException e) {
+                    throw new IllegalArgumentException(String
+                            .format("Skip sending the message because data URI scheme is invalid: %s", e.getMessage()));
                 }
             } else {
                 File file = new File(attachment);
@@ -264,14 +266,7 @@ public class PushoverMessageBuilder {
                     throw new IllegalArgumentException(
                             String.format("Skip sending the message as file '%s' does not exist.", attachment));
                 }
-                try {
-                    body.addFilePart(MESSAGE_KEY_ATTACHMENT, file.getName(),
-                            new PathContentProvider(contentType, file.toPath()), null);
-                } catch (IOException e) {
-                    logger.debug("IOException occurred - skip sending message: {}", e.getLocalizedMessage(), e);
-                    throw new PushoverCommunicationException(
-                            String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
-                }
+                addFilePart(file.toPath(), contentType);
             }
         }
 
@@ -282,5 +277,29 @@ public class PushoverMessageBuilder {
         }
 
         return body;
+    }
+
+    private Path createTempFile(byte[] data) throws PushoverCommunicationException {
+        try {
+            Path tmpFile = Files.createTempFile("pushover-", ".tmp");
+            return Files.write(tmpFile, data);
+        } catch (IOException e) {
+            logger.debug("IOException occurred while creating temp file - skip sending message: {}",
+                    e.getLocalizedMessage(), e);
+            throw new PushoverCommunicationException(
+                    String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+        }
+    }
+
+    private void addFilePart(Path path, @Nullable String contentType) throws PushoverCommunicationException {
+        try {
+            body.addFilePart(MESSAGE_KEY_ATTACHMENT, path.toFile().getName(),
+                    new PathContentProvider(contentType == null ? DEFAULT_CONTENT_TYPE : contentType, path), null);
+        } catch (IOException e) {
+            logger.debug("IOException occurred while adding content - skip sending message: {}",
+                    e.getLocalizedMessage(), e);
+            throw new PushoverCommunicationException(
+                    String.format("Skip sending the message: %s", e.getLocalizedMessage()), e);
+        }
     }
 }

--- a/bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/i18n/pushover.properties
+++ b/bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/i18n/pushover.properties
@@ -32,7 +32,7 @@ sendMonospaceMessageActionDescription = This method is used to send a monospace 
 sendAttachmentMessageActionLabel = send a plain text message with an attachment
 sendAttachmentMessageActionDescription = This method is used to send a message with an attachment.
 sendMessageActionInputAttachmentLabel = Attachment
-sendMessageActionInputAttachmentDescription = A (local) path to the attachment.
+sendMessageActionInputAttachmentDescription = Local path or URL to the attachment.
 sendMessageActionInputContentTypeLabel = Content Type
 sendMessageActionInputContentTypeDescription = The content type of the attachment. Defaults to "image/jpeg".
 

--- a/bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/i18n/pushover_de.properties
+++ b/bundles/org.openhab.binding.pushover/src/main/resources/OH-INF/i18n/pushover_de.properties
@@ -57,7 +57,7 @@ sendMonospaceMessageActionDescription = Action zum Versenden einer monospace-Nac
 sendAttachmentMessageActionLabel = eine Textnachricht mit Anhang senden
 sendAttachmentMessageActionDescription = Action zum Versenden einer Textnachricht mit Anhang.
 sendMessageActionInputAttachmentLabel = Anhang
-sendMessageActionInputAttachmentDescription = Lokaler Pfad zum Anhang.
+sendMessageActionInputAttachmentDescription = Lokaler Pfad oder URL zum Anhang.
 sendMessageActionInputContentTypeLabel = Content-Type
 sendMessageActionInputContentTypeDescription = Der Content-Type für den Anhang. Default: "image/jpeg".
 

